runtime/compiler/x/runtime/X86Codert.nasm

```diff
@@ -334,6 +334,72 @@ f2l_done:
 
 %endif ; TR_HOST_32BIT
 
+; crc32c
+;
+; Computes CRC32C checksum for a given buffer and size.
+;
+DECLARE_GLOBAL crc32c
+
+; rdi = crc > rax
+; rsi = bytes
+; rdx = off current rbx  - edx
+; rcx = end currently edx
+; tmp  = edi
+; tmp2 = r8
+align 16
+crc32c:
+        mov rax, rdi                        ; Copy current crc value into rax (result register)
+
+        begin512BitLoop:
+        lea rdi, [rdx+0x40]                 ;
+        cmp	rdi, rdx
+        jg begin64BitLoop                   ; Skip loop if less than 64 bytes in buffer
+
+        crc32 rax, qword [rsi+1*rdx+0x8]    ; 64-byte unrolled crc32c loop body
+        crc32 rax, qword [rsi+1*rdx+0x10]
+        crc32 rax, qword [rsi+1*rdx+0x18]
+        crc32 rax, qword [rsi+1*rdx+0x20]
+        crc32 rax, qword [rsi+1*rdx+0x28]
+        crc32 rax, qword [rsi+1*rdx+0x30]
+        crc32 rax, qword [rsi+1*rdx+0x38]
+        crc32 rax, qword [rsi+1*rdx+0x40]
+        add	rdx, 0x00000040                 ; Increase offset number of bytes written
+        jmp begin512BitLoop
+
+        begin64BitLoop:                     ;  crc 8 bytes at a time in loop
+        lea rdi, [rdx+0x8]
+        cmp	rdi, rcx
+        jg duff
+        crc32	rax, qword [rsi+1*rdx+0x8]
+        add	rdx, 0x00000008
+        jmp begin64BitLoop
+
+        duff:
+%define CRC32C_INC_INSTRUCTION_SIZE 10               ; size of crc32 + inc instruction
+        mov	rdi, rcx                                 ; Calculate where to jump to, depending on number of bytes left
+        sub	rdi, rdx
+        imul rdi, rdi, -CRC32C_INC_INSTRUCTION_SIZE
+        add	rdi, CRC32C_INC_INSTRUCTION_SIZE * 7
+        lea	r8, qword [rel crc7b]
+        add	rdi, r8
+        jmp	rdi                                      ; Jump to correct spot
+
+        crc7b:
+        crc32 rax, byte [rsi+1*rdx+0x8]
+        inc	rdx                                      ; Inc instruction required because num of bytes remaining is
+        crc32 rax, byte [rsi+1*rdx+0x8]              ; variable and the bytes must be processed in order
+        inc	rdx
+        crc32 rax, byte [rsi+1*rdx+0x8]
+        inc	rdx
+        crc32 rax, byte [rsi+1*rdx+0x8]
+        inc	rdx
+        crc32 rax, byte [rsi+1*rdx+0x8]
+        inc	rdx
+        crc32 rax, byte [rsi+1*rdx+0x8]
+        inc	rdx
+        crc32 rax, byte [rsi+1*rdx+0x8]
+        inc rdx
```

Comment:
Does this final `inc` need to be done?

EDIT: Oh, maybe it needs to be here if the "jmp rdi" can take you past the entire Duff's Device.  Is that true?

Reply:
Yes, there may not be an residual bytes to process